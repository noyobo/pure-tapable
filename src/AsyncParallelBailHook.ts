import { Async } from './Hook.js';
import type { DoneCallback, TapAsyncFn, TapPromiseFn } from './types';

/**
 * 并行串行 hook
 * 任一函数的返回非空值，则跳过剩余函数
 *
 * FIXME: 实现并行可中断
 */
export default class AsyncParallelBailHook<A extends unknown[]> extends Async<A, TapAsyncFn<A>, TapPromiseFn<A>> {
  callAsync = (...args: [...A, DoneCallback]) => {
    let taskIndex = 0;
    const asyncTaps = this.asyncTaps;
    const taskLength = asyncTaps.length;

    const done = args.pop() as DoneCallback;
    const params = this.argsToParams(args as unknown as A);

    const next = (error?: any, result?: any) => {
      if (error != undefined) return done(error);
      if (typeof result !== 'undefined') return done(null, result);

      if (taskIndex === taskLength) {
        done();
      } else {
        const { fn } = asyncTaps[taskIndex];
        fn.apply(this, [
          ...params,
          (error, result) => {
            taskIndex += 1;
            next(error, result);
          },
        ]);
      }
    };
    next();
  };

  promise = (...args: A): Promise<any> => {
    const params = this.argsToParams(args);
    let taskIndex = 0;

    return new Promise((resolve, reject) => {
      const next = async (value?: any) => {
        if (taskIndex === this.promiseTaps.length) {
          return resolve(value);
        }
        const { fn } = this.promiseTaps[taskIndex];
        return await fn(...params);
      };

      const checkDone = (result?: any) => {
        if (typeof result !== 'undefined') {
          return resolve(result);
        }
        next()
          .then((value) => {
            taskIndex += 1;
            checkDone(value);
          })
          .catch((error) => {
            reject(error);
          });
      };

      checkDone();
    });
  };
}
