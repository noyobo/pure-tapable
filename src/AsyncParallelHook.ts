import { Async } from './Hook.js';
import type { DoneCallback, TapAsyncFn, TapPromiseFn } from './types';

/**
 * 异步并行 hook
 * 忽略钩子函数的返回值
 */
export default class AsyncParallelHook<A extends unknown[]> extends Async<A, TapAsyncFn<A>, TapPromiseFn<A>> {
  callAsync = (...args: [...A, DoneCallback]) => {
    const done = args.pop() as DoneCallback;
    const argv = args as unknown as A;
    let isComplete = false;
    let taskDone = this.asyncTaps.length;

    // 所有异步函数执行完成触发 callAsync 回调
    const checkDone = (error: any) => {
      if (!isComplete) {
        taskDone -= 1;
        if (error != undefined) {
          isComplete = true;
          done(error);
        } else if (taskDone == 0) {
          isComplete = true;
          done();
        }
      }
    };

    for (let i = 0; i < this.asyncTaps.length; i++) {
      const { fn } = this.asyncTaps[i];
      // @ts-ignore
      fn.apply(this, [
        ...argv,
        function taskCallback(error) {
          checkDone(error);
        },
      ]);
    }
  };

  promise = (...args: A): Promise<void[]> => {
    return Promise.all(
      this.promiseTaps.map(({ fn }) => {
        return fn.apply(this, args);
      }),
    );
  };
}
