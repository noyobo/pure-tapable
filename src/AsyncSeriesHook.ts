import { Async } from './Hook.js';
import { DoneCallback, TapAsyncFn, TapPromiseFn } from './types';

/**
 * 异步串行 hook
 * 忽略钩子函数的返回值
 */
export default class AsyncSeriesHook<A extends unknown[]> extends Async<A, TapAsyncFn<A>, TapPromiseFn<A>> {
  callAsync = (...args: [...A, DoneCallback]) => {
    let taskIndex = 0;
    const asyncTaps = this.asyncTaps;
    const taskLength = asyncTaps.length;

    const done = args.pop() as DoneCallback;
    const params = this.argsToParams(args as unknown as A);

    const next = (error?: any, value?: any) => {
      if (error != undefined) return done(error);

      if (taskIndex === taskLength) {
        done();
      } else {
        const { fn } = asyncTaps[taskIndex];
        fn.apply(this, [
          ...params,
          (error, result) => {
            taskIndex += 1;
            next(error, result);
          },
        ]);
      }
    };
    next();
  };

  promise = (...args: A): Promise<any> => {
    const params = this.argsToParams(args);
    return new Promise((resolve, reject) => {
      let taskIndex = 0;
      const next = async (value?: any) => {
        if (taskIndex === this.promiseTaps.length) {
          return resolve(value);
        }

        const { fn } = this.promiseTaps[taskIndex];
        return await fn.apply(this, params);
      };

      const checkDone = () => {
        next()
          .then((value) => {
            taskIndex += 1;
            checkDone();
          })
          .catch((error) => {
            reject(error);
          });
      };

      checkDone();
    });
  };
}
