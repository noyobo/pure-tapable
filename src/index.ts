export { default as SyncHook } from './SyncHook';
export { default as SyncBailHook } from './SyncBailHook';
export { default as SyncWaterfallHook } from './SyncWaterfallHook';
export { default as SyncLoopHook } from './SyncLoopHook';

export { default as AsyncSeriesHook } from './AsyncSeriesHook';
export { default as AsyncSeriesBailHook } from './AsyncSeriesBailHook';
export { default as AsyncSeriesWaterfallHook } from './AsyncSeriesWaterfallHook';
export { default as AsyncParallelHook } from './AsyncParallelHook';
export { default as AsyncParallelBailHook } from './AsyncParallelBailHook';
