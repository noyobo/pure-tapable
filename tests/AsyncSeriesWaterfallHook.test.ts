import { AsyncSeriesWaterfallHook as S } from 'tapable';

import { AsyncSeriesWaterfallHook } from '../src';

describe('throwError', function () {
  it('throw error validate waterfall hooks args', () => {
    expect(() => {
      new AsyncSeriesWaterfallHook();
    }).toThrowError('Waterfall hooks must have at least one argument');
  });
});

describe('AsyncSeriesWaterfallHook callAsync', function () {
  it('multi', async function () {
    function run(Hook: typeof AsyncSeriesWaterfallHook, log: any[] = []): Promise<any[]> {
      return new Promise((resolve) => {
        const queue = new Hook<[string, number]>(['name', 'value']);
        queue.tapAsync('1', (name, value, cb) => {
          log.push([name, value, 1]);
          cb(null, 'after-1');
        });
        queue.tapAsync('2', (name, value, cb) => {
          log.push([name, value, 2]);
          cb(null, null);
        });
        queue.tapAsync('3', (name, value, cb) => {
          log.push([name, value, 3]);
          cb(null, 'after-3');
        });
        queue.callAsync('hooks', 0, (...args) => {
          log.push(['end']);
          log.push(args);
          resolve(log);
        });
      });
    }

    const a = await run(AsyncSeriesWaterfallHook);
    // @ts-ignore
    const b = await run(S);
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
    expect(a).toMatchObject(b);
  });

  it('multi callback void', async function () {
    function run(Hook: typeof AsyncSeriesWaterfallHook, log: any[] = []): Promise<any[]> {
      return new Promise((resolve) => {
        const queue = new Hook<[string, number]>(['name', 'value']);
        queue.tapAsync('1', (name, value, cb) => {
          log.push([name, value, 1]);
          cb();
        });
        queue.tapAsync('2', (name, value, cb) => {
          log.push([name, value, 2]);
          cb();
        });
        queue.tapAsync('3', (name, value, cb) => {
          log.push([name, value, 3]);
          cb();
        });
        queue.callAsync('hooks', 0, (...args) => {
          log.push(['end']);
          log.push(args);
          resolve(log);
        });
      });
    }

    const a = await run(AsyncSeriesWaterfallHook);
    // @ts-ignore
    const b = await run(S);
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
    expect(a).toMatchObject(b);
  });

  it('multi callback zero', async function () {
    function run(Hook: typeof AsyncSeriesWaterfallHook, log: any[] = []): Promise<any[]> {
      return new Promise((resolve) => {
        const queue = new Hook<[string, number]>(['name', 'value']);
        queue.tapAsync('1', (name, value, cb) => {
          log.push([name, value, 1]);
          cb();
        });
        queue.tapAsync('2', (name, value, cb) => {
          log.push([name, value, 2]);
          cb(null, 0);
        });
        queue.tapAsync('3', (name, value, cb) => {
          log.push([name, value, 3]);
          cb();
        });
        queue.callAsync('hooks', 0, (...args) => {
          log.push(['end']);
          log.push(args);
          resolve(log);
        });
      });
    }

    const a = await run(AsyncSeriesWaterfallHook);
    // @ts-ignore
    const b = await run(S);
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
    expect(a).toMatchObject(b);
  });

  it('multi callback error', async function () {
    function run(Hook: typeof AsyncSeriesWaterfallHook, log: any[] = []): Promise<any[]> {
      return new Promise((resolve) => {
        const queue = new Hook<[string, number]>(['name', 'value']);
        queue.tapAsync('1', (name, value, cb) => {
          log.push([name, value, 1]);
          cb(null, 'after-1');
        });
        queue.tapAsync('2', (name, value, cb) => {
          log.push([name, value, 2]);
          cb(new Error('error-2'));
        });
        queue.tapAsync('3', (name, value, cb) => {
          log.push([name, value, 3]);
          cb(null, 'after-3');
        });
        queue.callAsync('hooks', 0, (...args) => {
          log.push(['end']);
          log.push(args);
          resolve(log);
        });
      });
    }

    const a = await run(AsyncSeriesWaterfallHook);
    // @ts-ignore
    const b = await run(S);
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
    expect(a).toMatchObject(b);
  });

  it('single', async function () {
    function run(Hook: typeof AsyncSeriesWaterfallHook, log: any[] = []): Promise<any[]> {
      return new Promise((resolve) => {
        const queue = new Hook<[string]>(['name']);
        queue.tapAsync('1', (name, cb) => {
          log.push([name, 1]);
          cb(null, 'after-1');
        });
        queue.tapAsync('2', (name, cb) => {
          log.push([name, 2]);
          cb(null, 'after-2');
        });
        queue.tapAsync('3', (name, cb) => {
          log.push([name, 3]);
          cb(null, 'after-3');
        });
        queue.callAsync('hooks', (...args) => {
          log.push(['end']);
          log.push(args);
          resolve(log);
        });
      });
    }

    const a = await run(AsyncSeriesWaterfallHook);
    // @ts-ignore
    const b = await run(S);
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
    expect(a).toMatchObject(b);
  });

  it('single callback void', async function () {
    function run(Hook: typeof AsyncSeriesWaterfallHook, log: any[] = []): Promise<any[]> {
      return new Promise((resolve) => {
        const queue = new Hook<[string]>(['name']);
        queue.tapAsync('1', (name, cb) => {
          log.push([name, 1]);
          cb();
        });
        queue.tapAsync('2', (name, cb) => {
          log.push([name, 2]);
          cb();
        });
        queue.tapAsync('3', (name, cb) => {
          log.push([name, 3]);
          cb();
        });
        queue.callAsync('hooks', (...args) => {
          log.push(['end']);
          log.push(args);
          resolve(log);
        });
      });
    }

    const a = await run(AsyncSeriesWaterfallHook);
    // @ts-ignore
    const b = await run(S);
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
    expect(a).toMatchObject(b);
  });

  it('single callback false', async function () {
    function run(Hook: typeof AsyncSeriesWaterfallHook, log: any[] = []): Promise<any[]> {
      return new Promise((resolve) => {
        const queue = new Hook<[string]>(['name']);
        queue.tapAsync('1', (name, cb) => {
          log.push([name, 1]);
          cb();
        });
        queue.tapAsync('2', (name, cb) => {
          log.push([name, 2]);
          cb(null, false);
        });
        queue.tapAsync('3', (name, cb) => {
          log.push([name, 3]);
          cb();
        });
        queue.callAsync('hooks', (...args) => {
          log.push(['end']);
          log.push(args);
          resolve(log);
        });
      });
    }

    const a = await run(AsyncSeriesWaterfallHook);
    // @ts-ignore
    const b = await run(S);
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
    expect(a).toMatchObject(b);
  });

  it('single callback error', async function () {
    function run(Hook: typeof AsyncSeriesWaterfallHook, log: any[] = []): Promise<any[]> {
      return new Promise((resolve) => {
        const queue = new Hook<[string]>(['name']);
        queue.tapAsync('1', (name, cb) => {
          log.push([name, 1]);
          cb(null, 'after-1');
        });
        queue.tapAsync('2', (name, cb) => {
          log.push([name, 2]);
          cb(new Error('error'));
        });
        queue.tapAsync('3', (name, cb) => {
          log.push([name, 3]);
          cb(null, 'after-3');
        });
        queue.callAsync('hooks', (...args) => {
          log.push(['end']);
          log.push(args);
          resolve(log);
        });
      });
    }

    const a = await run(AsyncSeriesWaterfallHook);
    // @ts-ignore
    const b = await run(S);
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
    expect(a).toMatchObject(b);
  });
});

describe('AsyncSeriesWaterfallHook promise', function () {
  it('multi', async function () {
    function run(Hook: typeof AsyncSeriesWaterfallHook, log: any[] = []): Promise<any[]> {
      return new Promise((resolve) => {
        const queue = new Hook<[string, number]>(['name', 'value']);
        queue.tapPromise('1', (name, value) => {
          return new Promise((resolve) => {
            log.push([name, value, 1]);
            resolve('after-1');
          });
        });
        queue.tapPromise('2', (name, value) => {
          return new Promise((resolve) => {
            log.push([name, value, 2]);
            resolve(null);
          });
        });
        queue.tapPromise('3', (name, value) => {
          return new Promise((resolve) => {
            log.push([name, value, 3]);
            resolve('after-3');
          });
        });
        queue.promise('hooks', 0).then((...args) => {
          log.push(['end']);
          log.push(args);
          resolve(log);
        });
      });
    }

    const a = await run(AsyncSeriesWaterfallHook);
    // @ts-ignore
    const b = await run(S);
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
    expect(a).toMatchObject(b);
  });

  it('multi resolve void', async function () {
    function run(Hook: typeof AsyncSeriesWaterfallHook, log: any[] = []): Promise<any[]> {
      return new Promise((resolve) => {
        const queue = new Hook<[string, number]>(['name', 'value']);
        queue.tapPromise('1', (name, value) => {
          return new Promise((resolve) => {
            log.push([name, value, 1]);
            resolve(undefined);
          });
        });
        queue.tapPromise('2', (name, value) => {
          return new Promise((resolve) => {
            log.push([name, value, 2]);
            resolve(undefined);
          });
        });
        queue.tapPromise('3', (name, value) => {
          return new Promise((resolve) => {
            log.push([name, value, 3]);
            resolve(undefined);
          });
        });
        queue.promise('hooks', 0).then((...args) => {
          log.push(['end']);
          log.push(args);
          resolve(log);
        });
      });
    }

    const a = await run(AsyncSeriesWaterfallHook);
    // @ts-ignore
    const b = await run(S);
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
    expect(a).toMatchObject(b);
  });

  it('multi resolve zero', async function () {
    function run(Hook: typeof AsyncSeriesWaterfallHook, log: any[] = []): Promise<any[]> {
      return new Promise((resolve) => {
        const queue = new Hook<[string, number]>(['name', 'value']);
        queue.tapPromise('1', (name, value) => {
          return new Promise((resolve) => {
            log.push([name, value, 1]);
            resolve(undefined);
          });
        });
        queue.tapPromise('2', (name, value) => {
          return new Promise((resolve) => {
            log.push([name, value, 2]);
            resolve(0);
          });
        });
        queue.tapPromise('3', (name, value) => {
          return new Promise((resolve) => {
            log.push([name, value, 3]);
            resolve(undefined);
          });
        });
        queue.promise('hooks', 0).then((...args) => {
          log.push(['end']);
          log.push(args);
          resolve(log);
        });
      });
    }

    const a = await run(AsyncSeriesWaterfallHook);
    // @ts-ignore
    const b = await run(S);
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
    expect(a).toMatchObject(b);
  });

  it('multi resolve error', async function () {
    function run(Hook: typeof AsyncSeriesWaterfallHook, log: any[] = []): Promise<any[]> {
      return new Promise((resolve) => {
        const queue = new Hook<[string, number]>(['name', 'value']);
        queue.tapPromise('1', (name, value) => {
          return new Promise((resolve) => {
            log.push([name, value, 1]);
            resolve(undefined);
          });
        });
        queue.tapPromise('2', (name, value) => {
          return new Promise((resolve) => {
            log.push([name, value, 2]);
            resolve(new Error('error'));
          });
        });
        queue.tapPromise('3', (name, value) => {
          return new Promise((resolve) => {
            log.push([name, value, 3]);
            resolve(undefined);
          });
        });
        queue.promise('hooks', 0).then((...args) => {
          log.push(['end']);
          log.push(args);
          resolve(log);
        });
      });
    }

    const a = await run(AsyncSeriesWaterfallHook);
    // @ts-ignore
    const b = await run(S);
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
    expect(a).toMatchObject(b);
  });

  it('multi reject error', async function () {
    function run(Hook: typeof AsyncSeriesWaterfallHook, log: any[] = []): Promise<any[]> {
      return new Promise((resolve) => {
        const queue = new Hook<[string, number]>(['name', 'value']);
        queue.tapPromise('1', (name, value) => {
          return new Promise((resolve) => {
            log.push([name, value, 1]);
            resolve(undefined);
          });
        });
        queue.tapPromise('2', (name, value) => {
          return new Promise((resolve, reject) => {
            log.push([name, value, 2]);
            reject(new Error('error'));
          });
        });
        queue.tapPromise('3', (name, value) => {
          return new Promise((resolve) => {
            log.push([name, value, 3]);
            resolve(undefined);
          });
        });
        queue.promise('hooks', 0).catch((...args) => {
          log.push(['end']);
          log.push(args);
          resolve(log);
        });
      });
    }

    const a = await run(AsyncSeriesWaterfallHook);
    // @ts-ignore
    const b = await run(S);
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
    expect(a).toMatchObject(b);
  });

  it('single', async function () {
    function run(Hook: typeof AsyncSeriesWaterfallHook, log: any[] = []): Promise<any[]> {
      return new Promise((resolve) => {
        const queue = new Hook<[string]>(['name']);
        queue.tapPromise('1', (name) => {
          return new Promise((resolve) => {
            log.push([name, 1]);
            resolve('after-1');
          });
        });
        queue.tapPromise('2', (name) => {
          return new Promise((resolve) => {
            log.push([name, 2]);
            resolve('after-2');
          });
        });
        queue.tapPromise('3', (name) => {
          return new Promise((resolve) => {
            log.push([name, 3]);
            resolve('after-3');
          });
        });
        queue.promise('hooks').then((...args) => {
          log.push(['end']);
          log.push(args);
          resolve(log);
        });
      });
    }

    const a = await run(AsyncSeriesWaterfallHook);
    // @ts-ignore
    const b = await run(S);
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
    expect(a).toMatchObject(b);
  });

  it('single resolve void', async function () {
    function run(Hook: typeof AsyncSeriesWaterfallHook, log: any[] = []): Promise<any[]> {
      return new Promise((resolve) => {
        const queue = new Hook<[string]>(['name']);
        queue.tapPromise('1', (name) => {
          return new Promise((resolve) => {
            log.push([name, 1]);
            resolve(undefined);
          });
        });
        queue.tapPromise('2', (name) => {
          return new Promise((resolve) => {
            log.push([name, 2]);
            resolve(undefined);
          });
        });
        queue.tapPromise('3', (name) => {
          return new Promise((resolve) => {
            log.push([name, 3]);
            resolve(undefined);
          });
        });
        queue.promise('hooks').then((...args) => {
          log.push(['end']);
          log.push(args);
          resolve(log);
        });
      });
    }

    const a = await run(AsyncSeriesWaterfallHook);
    // @ts-ignore
    const b = await run(S);
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
    expect(a).toMatchObject(b);
  });

  it('single resolve zero', async function () {
    function run(Hook: typeof AsyncSeriesWaterfallHook, log: any[] = []): Promise<any[]> {
      return new Promise((resolve) => {
        const queue = new Hook<[string]>(['name', 'value']);
        queue.tapPromise('1', (name) => {
          return new Promise((resolve) => {
            log.push([name, 1]);
            resolve(undefined);
          });
        });
        queue.tapPromise('2', (name) => {
          return new Promise((resolve) => {
            log.push([name, 2]);
            resolve(0);
          });
        });
        queue.tapPromise('3', (name) => {
          return new Promise((resolve) => {
            log.push([name, 3]);
            resolve(undefined);
          });
        });
        queue.promise('hooks').then((...args) => {
          log.push(['end']);
          log.push(args);
          resolve(log);
        });
      });
    }

    const a = await run(AsyncSeriesWaterfallHook);
    // @ts-ignore
    const b = await run(S);
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
    expect(a).toMatchObject(b);
  });

  it('single resolve error', async function () {
    function run(Hook: typeof AsyncSeriesWaterfallHook, log: any[] = []): Promise<any[]> {
      return new Promise((resolve) => {
        const queue = new Hook<[string]>(['name']);
        queue.tapPromise('1', (name) => {
          return new Promise((resolve) => {
            log.push([name, 1]);
            resolve(undefined);
          });
        });
        queue.tapPromise('2', (name) => {
          return new Promise((resolve) => {
            log.push([name, 2]);
            resolve(new Error('error'));
          });
        });
        queue.tapPromise('3', (name) => {
          return new Promise((resolve) => {
            log.push([name, 3]);
            resolve(undefined);
          });
        });
        queue.promise('hooks').then((...args) => {
          log.push(['end']);
          log.push(args);
          resolve(log);
        });
      });
    }

    const a = await run(AsyncSeriesWaterfallHook);
    // @ts-ignore
    const b = await run(S);
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
    expect(a).toMatchObject(b);
  });

  it('single reject error', async function () {
    function run(Hook: typeof AsyncSeriesWaterfallHook, log: any[] = []): Promise<any[]> {
      return new Promise((resolve) => {
        const queue = new Hook<[string]>(['name']);
        queue.tapPromise('1', (name) => {
          return new Promise((resolve) => {
            log.push([name, 1]);
            resolve(undefined);
          });
        });
        queue.tapPromise('2', (name) => {
          return new Promise((resolve, reject) => {
            log.push([name, 2]);
            reject(new Error('error'));
          });
        });
        queue.tapPromise('3', (name) => {
          return new Promise((resolve) => {
            log.push([name, 3]);
            resolve(undefined);
          });
        });
        queue.promise('hooks').catch((...args) => {
          log.push(['end']);
          log.push(args);
          resolve(log);
        });
      });
    }

    const a = await run(AsyncSeriesWaterfallHook);
    // @ts-ignore
    const b = await run(S);
    expect(a).toMatchSnapshot();
    expect(b).toMatchSnapshot();
    expect(a).toMatchObject(b);
  });
});
