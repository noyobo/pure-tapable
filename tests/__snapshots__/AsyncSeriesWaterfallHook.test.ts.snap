// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AsyncSeriesWaterfallHook callAsync multi 1`] = `
[
  [
    "hooks",
    0,
    1,
  ],
  [
    "after-1",
    0,
    2,
  ],
  [
    null,
    0,
    3,
  ],
  [
    "end",
  ],
  [
    null,
    "after-3",
  ],
]
`;

exports[`AsyncSeriesWaterfallHook callAsync multi 2`] = `
[
  [
    "hooks",
    0,
    1,
  ],
  [
    "after-1",
    0,
    2,
  ],
  [
    null,
    0,
    3,
  ],
  [
    "end",
  ],
  [
    null,
    "after-3",
  ],
]
`;

exports[`AsyncSeriesWaterfallHook callAsync multi callback error 1`] = `
[
  [
    "hooks",
    0,
    1,
  ],
  [
    "after-1",
    0,
    2,
  ],
  [
    "end",
  ],
  [
    [Error: error-2],
  ],
]
`;

exports[`AsyncSeriesWaterfallHook callAsync multi callback error 2`] = `
[
  [
    "hooks",
    0,
    1,
  ],
  [
    "after-1",
    0,
    2,
  ],
  [
    "end",
  ],
  [
    [Error: error-2],
  ],
]
`;

exports[`AsyncSeriesWaterfallHook callAsync multi callback void 1`] = `
[
  [
    "hooks",
    0,
    1,
  ],
  [
    "hooks",
    0,
    2,
  ],
  [
    "hooks",
    0,
    3,
  ],
  [
    "end",
  ],
  [
    null,
    "hooks",
  ],
]
`;

exports[`AsyncSeriesWaterfallHook callAsync multi callback void 2`] = `
[
  [
    "hooks",
    0,
    1,
  ],
  [
    "hooks",
    0,
    2,
  ],
  [
    "hooks",
    0,
    3,
  ],
  [
    "end",
  ],
  [
    null,
    "hooks",
  ],
]
`;

exports[`AsyncSeriesWaterfallHook callAsync multi callback zero 1`] = `
[
  [
    "hooks",
    0,
    1,
  ],
  [
    "hooks",
    0,
    2,
  ],
  [
    0,
    0,
    3,
  ],
  [
    "end",
  ],
  [
    null,
    0,
  ],
]
`;

exports[`AsyncSeriesWaterfallHook callAsync multi callback zero 2`] = `
[
  [
    "hooks",
    0,
    1,
  ],
  [
    "hooks",
    0,
    2,
  ],
  [
    0,
    0,
    3,
  ],
  [
    "end",
  ],
  [
    null,
    0,
  ],
]
`;

exports[`AsyncSeriesWaterfallHook callAsync single 1`] = `
[
  [
    "hooks",
    1,
  ],
  [
    "after-1",
    2,
  ],
  [
    "after-2",
    3,
  ],
  [
    "end",
  ],
  [
    null,
    "after-3",
  ],
]
`;

exports[`AsyncSeriesWaterfallHook callAsync single 2`] = `
[
  [
    "hooks",
    1,
  ],
  [
    "after-1",
    2,
  ],
  [
    "after-2",
    3,
  ],
  [
    "end",
  ],
  [
    null,
    "after-3",
  ],
]
`;

exports[`AsyncSeriesWaterfallHook callAsync single callback error 1`] = `
[
  [
    "hooks",
    1,
  ],
  [
    "after-1",
    2,
  ],
  [
    "end",
  ],
  [
    [Error: error],
  ],
]
`;

exports[`AsyncSeriesWaterfallHook callAsync single callback error 2`] = `
[
  [
    "hooks",
    1,
  ],
  [
    "after-1",
    2,
  ],
  [
    "end",
  ],
  [
    [Error: error],
  ],
]
`;

exports[`AsyncSeriesWaterfallHook callAsync single callback false 1`] = `
[
  [
    "hooks",
    1,
  ],
  [
    "hooks",
    2,
  ],
  [
    false,
    3,
  ],
  [
    "end",
  ],
  [
    null,
    false,
  ],
]
`;

exports[`AsyncSeriesWaterfallHook callAsync single callback false 2`] = `
[
  [
    "hooks",
    1,
  ],
  [
    "hooks",
    2,
  ],
  [
    false,
    3,
  ],
  [
    "end",
  ],
  [
    null,
    false,
  ],
]
`;

exports[`AsyncSeriesWaterfallHook callAsync single callback void 1`] = `
[
  [
    "hooks",
    1,
  ],
  [
    "hooks",
    2,
  ],
  [
    "hooks",
    3,
  ],
  [
    "end",
  ],
  [
    null,
    "hooks",
  ],
]
`;

exports[`AsyncSeriesWaterfallHook callAsync single callback void 2`] = `
[
  [
    "hooks",
    1,
  ],
  [
    "hooks",
    2,
  ],
  [
    "hooks",
    3,
  ],
  [
    "end",
  ],
  [
    null,
    "hooks",
  ],
]
`;

exports[`AsyncSeriesWaterfallHook promise multi 1`] = `
[
  [
    "hooks",
    0,
    1,
  ],
  [
    "after-1",
    0,
    2,
  ],
  [
    null,
    0,
    3,
  ],
  [
    "end",
  ],
  [
    "after-3",
  ],
]
`;

exports[`AsyncSeriesWaterfallHook promise multi 2`] = `
[
  [
    "hooks",
    0,
    1,
  ],
  [
    "after-1",
    0,
    2,
  ],
  [
    null,
    0,
    3,
  ],
  [
    "end",
  ],
  [
    "after-3",
  ],
]
`;

exports[`AsyncSeriesWaterfallHook promise multi reject error 1`] = `
[
  [
    "hooks",
    0,
    1,
  ],
  [
    "hooks",
    0,
    2,
  ],
  [
    "end",
  ],
  [
    [Error: error],
  ],
]
`;

exports[`AsyncSeriesWaterfallHook promise multi reject error 2`] = `
[
  [
    "hooks",
    0,
    1,
  ],
  [
    "hooks",
    0,
    2,
  ],
  [
    "end",
  ],
  [
    [Error: error],
  ],
]
`;

exports[`AsyncSeriesWaterfallHook promise multi resolve error 1`] = `
[
  [
    "hooks",
    0,
    1,
  ],
  [
    "hooks",
    0,
    2,
  ],
  [
    [Error: error],
    0,
    3,
  ],
  [
    "end",
  ],
  [
    [Error: error],
  ],
]
`;

exports[`AsyncSeriesWaterfallHook promise multi resolve error 2`] = `
[
  [
    "hooks",
    0,
    1,
  ],
  [
    "hooks",
    0,
    2,
  ],
  [
    [Error: error],
    0,
    3,
  ],
  [
    "end",
  ],
  [
    [Error: error],
  ],
]
`;

exports[`AsyncSeriesWaterfallHook promise multi resolve void 1`] = `
[
  [
    "hooks",
    0,
    1,
  ],
  [
    "hooks",
    0,
    2,
  ],
  [
    "hooks",
    0,
    3,
  ],
  [
    "end",
  ],
  [
    "hooks",
  ],
]
`;

exports[`AsyncSeriesWaterfallHook promise multi resolve void 2`] = `
[
  [
    "hooks",
    0,
    1,
  ],
  [
    "hooks",
    0,
    2,
  ],
  [
    "hooks",
    0,
    3,
  ],
  [
    "end",
  ],
  [
    "hooks",
  ],
]
`;

exports[`AsyncSeriesWaterfallHook promise multi resolve zero 1`] = `
[
  [
    "hooks",
    0,
    1,
  ],
  [
    "hooks",
    0,
    2,
  ],
  [
    0,
    0,
    3,
  ],
  [
    "end",
  ],
  [
    0,
  ],
]
`;

exports[`AsyncSeriesWaterfallHook promise multi resolve zero 2`] = `
[
  [
    "hooks",
    0,
    1,
  ],
  [
    "hooks",
    0,
    2,
  ],
  [
    0,
    0,
    3,
  ],
  [
    "end",
  ],
  [
    0,
  ],
]
`;

exports[`AsyncSeriesWaterfallHook promise single 1`] = `
[
  [
    "hooks",
    1,
  ],
  [
    "after-1",
    2,
  ],
  [
    "after-2",
    3,
  ],
  [
    "end",
  ],
  [
    "after-3",
  ],
]
`;

exports[`AsyncSeriesWaterfallHook promise single 2`] = `
[
  [
    "hooks",
    1,
  ],
  [
    "after-1",
    2,
  ],
  [
    "after-2",
    3,
  ],
  [
    "end",
  ],
  [
    "after-3",
  ],
]
`;

exports[`AsyncSeriesWaterfallHook promise single reject error 1`] = `
[
  [
    "hooks",
    1,
  ],
  [
    "hooks",
    2,
  ],
  [
    "end",
  ],
  [
    [Error: error],
  ],
]
`;

exports[`AsyncSeriesWaterfallHook promise single reject error 2`] = `
[
  [
    "hooks",
    1,
  ],
  [
    "hooks",
    2,
  ],
  [
    "end",
  ],
  [
    [Error: error],
  ],
]
`;

exports[`AsyncSeriesWaterfallHook promise single resolve error 1`] = `
[
  [
    "hooks",
    1,
  ],
  [
    "hooks",
    2,
  ],
  [
    [Error: error],
    3,
  ],
  [
    "end",
  ],
  [
    [Error: error],
  ],
]
`;

exports[`AsyncSeriesWaterfallHook promise single resolve error 2`] = `
[
  [
    "hooks",
    1,
  ],
  [
    "hooks",
    2,
  ],
  [
    [Error: error],
    3,
  ],
  [
    "end",
  ],
  [
    [Error: error],
  ],
]
`;

exports[`AsyncSeriesWaterfallHook promise single resolve void 1`] = `
[
  [
    "hooks",
    1,
  ],
  [
    "hooks",
    2,
  ],
  [
    "hooks",
    3,
  ],
  [
    "end",
  ],
  [
    "hooks",
  ],
]
`;

exports[`AsyncSeriesWaterfallHook promise single resolve void 2`] = `
[
  [
    "hooks",
    1,
  ],
  [
    "hooks",
    2,
  ],
  [
    "hooks",
    3,
  ],
  [
    "end",
  ],
  [
    "hooks",
  ],
]
`;

exports[`AsyncSeriesWaterfallHook promise single resolve zero 1`] = `
[
  [
    "hooks",
    1,
  ],
  [
    "hooks",
    2,
  ],
  [
    0,
    3,
  ],
  [
    "end",
  ],
  [
    0,
  ],
]
`;

exports[`AsyncSeriesWaterfallHook promise single resolve zero 2`] = `
[
  [
    "hooks",
    1,
  ],
  [
    "hooks",
    2,
  ],
  [
    0,
    3,
  ],
  [
    "end",
  ],
  [
    0,
  ],
]
`;
